lib/include/modelSpec.h

```diff
@@ -436,6 +437,114 @@ class NNmodel
+            gennError("You need to call initGeNN first.");
+        }
+        if (final) {
+            gennError("Trying to add a current source to a finalized model.");
+        }
+
+#ifdef MPI_ENABLE
+        // Determine the host ID
+        int mpiHostID = 0;
+        MPI_Comm_rank(MPI_COMM_WORLD, &mpiHostID);
+
+        // Pick map to add group to appropriately
+        auto &groupMap = (hostID == mpiHostID) ? m_LocalCurrentSources : m_RemoteCurrentSources;
+#else
+        // If MPI is disabled always add to local neuron groups and zero host id
+        auto &groupMap = m_LocalCurrentSources;
+        hostID = 0;
+#endif
+
+        // Add neuron group to map
+        auto result = groupMap.emplace(std::piecewise_construct,
+            std::forward_as_tuple(name),
+            std::forward_as_tuple(name, model,
+                                  paramValues.getValues(), varInitialisers.getInitialisers(), hostID, deviceID));
+
+        if(!result.second)
+        {
+            gennError("Cannot add a current source with duplicate name:" + name);
+            return NULL;
+        }
+        else
+        {
+            return &result.first->second;
+        }
+    }
+
+    //! Adds a new current source to the model using a singleton current source model created using standard DECLARE_MODEL and IMPLEMENT_MODEL macros
+    /*! \tparam CurrentSourceModel type of neuron model (derived from CurrentSourceModel::Base).
+        \param name string containing unique name of current source.
+        \param paramValues parameters for model wrapped in CurrentSourceModel::ParamValues object.
+        \param varInitialisers state variable initialiser snippets and parameters wrapped in CurrentSourceModel::VarValues object.
+        \return pointer to newly created CurrentSource */
+    template<typename CurrentSourceModel>
+    CurrentSource *addCurrentSource(const string &name,
+                                     const typename CurrentSourceModel::ParamValues &paramValues,
+                                     const typename CurrentSourceModel::VarValues &varInitialisers,
+                                     int hostID = 0, int deviceID = 0)
+    {
+        return addCurrentSource<CurrentSourceModel>(name, CurrentSourceModel::getInstance(), paramValues, varInitialisers, hostID, deviceID);
+    }
+
+    //! Injects an existing current source into an existing neuron group
+    /*! \param currentSourceName string name of the current source
+        \param targetNeuronGroupName string name of the targen neuron group */
+    void injectCurrent(const string &currentSourceName, const string &targetNeuronGroupName)
+    {
+        auto targetGroup = findNeuronGroup(targetNeuronGroupName);
+        auto source = findCurrentSource(currentSourceName);
+        targetGroup->injectCurrent(source);
+    }
+    
+
+    // PUBLIC STATIC FUNCTIONS
+    //========================
+    
+    static void allocateExtraGlobalParam( void** egp, int size )
```

Comment:
What do these do? There shouldn't really be any CUDA-specific stuff in ``NNmodel``

Reply:
Oops they don't really belong to this commit. I added them after the conversation we had about `SpikeSourceArray` where you have mentioned that it would be good to have functions to allocate and free `extraGlobalParam`s

Reply:
Ok cool - if you can remove them from this branch,  we can look at this again when it comes to merging the python_wrapper branch